processor/spanmetricsprocessor/processor.go

```diff
@@ -381,7 +381,13 @@ func (p *processorImp) aggregateMetricsForServiceSpans(rspans ptrace.ResourceSpa
 }
 
 func (p *processorImp) aggregateMetricsForSpan(serviceName string, span ptrace.Span, resourceAttr pcommon.Map) {
-	latencyInMilliseconds := float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
+	// Protect against end timestamps before start timestamps. Assume 0 duration.
+	latencyInMilliseconds := float64(0)
```

Comment:
It seems like doing a math.Max with 0 and the calculated value would be more straightforward & easier to understand for future maintainers.

Reply:
@dehaansa Part of the issue is that the timestamps are uint64, so the result of subtraction is always going to be a positive integer.  math.Max will not solve the issue in that case.

Reply:
> Part of the issue is that the timestamps are uint64, so the result of subtraction is always going to be a positive integer. math.Max will not solve the issue in that case.

This doesn't make sense.

This calculation: 

```go
latencyInMilliseconds := float64(span.EndTimestamp()-span.StartTimestamp()) / float64(time.Millisecond.Nanoseconds())
```

will result in a `float64` that is `>0` if the end time is after the start time and `<0` if it is before the start time.  So, in that case can't we follow up with `math.Max()` since we're now in `float64` land?

```go
latencyInMilliseconds = math.Max(0, latencyInMilliseconds)
```

We could rewrite this alternately:

```go
max := func(a, b int) int {
  if a > b {
    return a
  }
  return b
}

intDelta := span.EndTimestamp()-span.SpartTimestamp()
clampedDelta := max(0, intDelta)
latencyInMs := float64(clampedDelta) / float64(time.Millisecond)
// time.Millisecond is already an integer number of nanoseconds
// same as clampedDelta
```

Reply:
@Aneurysm9 Let me clarify,

If you have a negative result from subtraction, using uint64 would mean the sign bit is ignored, resulting in a large positive number instead of a negative number.  When you attempt the clamp, the value is already going to be positive.

You can see this with: https://go.dev/play/p/mHQ3_zjFzlq

This is why we need to validate that end time is after start time instead.

Reply:
@Aneurysm9 @dehaansa @bogdandrutu I would really like to get this in, its been over a month and a half since I provided this solution.  The current implementation crashes the collector anytime problematic spans are provided.

Reply:
Treating timestamps as `uint64` is a mistake and can't be assumed to be honored by all languages.  [The spec anticipates](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#time) that timestamps will be represented using language-preferred mechanisms, which in Go is `time.Time`, aka `int64` nanoseconds from the UNIX epoch.  The [proto uses `fixed64`](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/trace/v1/trace.proto#L164), which is presumably why `pdata` uses `uint64`, but even `pdata` [assumes that values can be cast](https://github.com/open-telemetry/opentelemetry-collector/blob/734e25e8dec549c4a408f66b9ebd445e52f539fb/pdata/internal/timestamp.go#L32) to `int64` sanely.

If you use `span.EndTimestamp().AsTime()` you'll get a `time.Time` value which [will work correctly](https://go.dev/play/p/M1rxVBlg8-z).

Reply:
I agree that treating the timestamps as a `uint64` is a mistake generally, however, this is golang and the current underlying implementation for this data is `uint64`. The calculation is only in this specific scope geared towards incrementing a specific bucket and isn't bleeding out of this function.

Both methods will result in the same effect.  Converting to timestamps and using Sub function has additional overhead associated with it that a subtraction and if statement doesn't.

I'm not against doing the switch, it's pretty simple, but is it worth the additional overhead for a throwaway calculation to be the correct data type?

Reply:
pinging @Aneurysm9

Reply:
If I'm understanding things correctly:
1. This PR resolves a user-facing correctness problem.
2. Whether or not `pdata.Timestamp` should be `uint64` is a question of code quality / quality of life. It applies to a larger context than this PR.

If so, I suggest this PR should be merged now, and a separate proposal can be made to discuss the `uint64` question.